include/kernels/staggered_kd_build_xinv_kernel.cuh

```diff
@@ -0,0 +1,140 @@
+    int coord[nDim];
+    int coord_coarse[nDim];
+
+    getCoords(coord, x_cb, arg.x_size, parity);
+
+    // Compute coarse coordinates
+    for (int d = 0; d < nDim; d++) coord_coarse[d] = coord[d]/2;
+    int coarse_parity = 0;
+    for (int d = 0; d < nDim; d++) coarse_parity += coord_coarse[d];
+    coarse_parity &= 1;
+    int coarse_x_cb = ((coord_coarse[3]*arg.xc_size[2]+coord_coarse[2])*arg.xc_size[1]+coord_coarse[1])*(arg.xc_size[0]/2) + coord_coarse[0]/2;
+
+    // Fine parity gives the coarse spin
+    constexpr int s = 0; // fine spin is always 0, since it's staggered.
+    const int s_c_row = arg.spin_map(s,parity); // Coarse spin row index
+    const int s_c_col = arg.spin_map(s,1-parity); // Coarse spin col index
+
+    // The coarse color row depends on my fine hypercube corner.
+    int hyperCorner = 4*(coord[3]%2)+2*(coord[2]%2)+(coord[1]%2);
+    int c_row = 8*ic_f + hyperCorner;
+
+#pragma unroll
+    for (int mu=0; mu < nDim; mu++) {
+      // The coarse color column depends on my fine hypercube+mu corner.
+      coord[mu]++; // linkIndexP1, it's fine if this wraps because we're modding by 2.
+      int hyperCorner_mu = 4*(coord[3]%2)+2*(coord[2]%2)+(coord[1]%2);
+      coord[mu]--;
+
+      int c_col = 8*jc_f + hyperCorner_mu;
+
+      //Check to see if we are on the edge of a block.  If adjacent site
+      //is in same block, M = X, else it's a hopping term and we ignore it
+      const bool isDiagonal = ((coord[mu]+1)%arg.x_size[mu])/2 == coord_coarse[mu] ? true : false;
+
+      complex<Float> vuv = arg.U(mu,parity,x_cb,ic_f,jc_f);
+
+      if (isDiagonal) {
+        // backwards
+        arg.X(0,coarse_parity,coarse_x_cb,s_c_col,s_c_row,c_col,c_row) = conj(vuv);
+        // forwards
+        arg.X(0,coarse_parity,coarse_x_cb,s_c_row,s_c_col,c_row,c_col) = -vuv;
+      } // end (isDiagonal)
+    }
+
+    // add staggered mass term to diagonal
+    if (ic_f == 0 && jc_f == 0 && x_cb < arg.coarseVolumeCB) {
+#pragma unroll
+      for (int s = 0; s < coarseSpin; s++) {
+#pragma unroll
+        for (int c = 0; c < coarseColor; c++) {
+          arg.X(0,parity,x_cb,s,s,c,c) = static_cast<Float>(2.0) * complex<Float>(arg.mass,0.0); // staggered conventions. No need to +=
+        } //Color
+      } //Spin
+    }
+  }
+
+  template<typename Float, int fineColor, int coarseSpin, int coarseColor, typename Arg>
+  void ComputeStaggeredKDBlockCPU(Arg arg)
+  {
+    for (int parity=0; parity<2; parity++) {
+#pragma omp parallel for
+      for (int x_cb=0; x_cb<arg.fineVolumeCB; x_cb++) { // Loop over fine volume
+        for (int ic_f=0; ic_f<fineColor; ic_f++) {
+          for (int jc_f=0; jc_f<fineColor; jc_f++) {
+            ComputeStaggeredKDBlock<Float,fineColor,coarseSpin,coarseColor>(arg, parity, x_cb, ic_f, jc_f);
+          } // coarse color columns
+        } // coarse color rows
+      } // c/b volume
+    } // parity
+  }
+
+  template<typename Float, int fineColor, int coarseSpin, int coarseColor, typename Arg>
```

Comment:
For minimal changes for generic kernels, please remove all the template parameters except `Arg` and make all the other template parameters members of `Arg`

Reply:
Done